// ng-number-picker/esm5/lib/number-picker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter } from '@angular/core';
import { NumberPickerService } from './number-picker.service';
var NumberPickerComponent = /** @class */ (function () {
    function NumberPickerComponent(numberPickerService) {
        this.numberPickerService = numberPickerService;
        this.eventHolder = null;
        this.countInterval = null;
        this.isInputFocused = false;
        this.size = 'md';
        this.customClass = {};
        this.mouseWheel = false;
        this.arrowKeys = true;
        this.inputReadOnly = false;
        this.showUpButton = true;
        this.showDownButton = true;
        this.valueChange = new EventEmitter();
        this.minReached = new EventEmitter();
        this.maxReached = new EventEmitter();
        this.pickStarted = new EventEmitter();
        this.pickStoped = new EventEmitter();
        this.pickUpStarted = new EventEmitter();
        this.pickUpStoped = new EventEmitter();
        this.pickDownStarted = new EventEmitter();
        this.pickDownStoped = new EventEmitter();
    }
    /**
     * @return {?}
     */
    NumberPickerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.initPicker();
    };
    /**
     * @return {?}
     */
    NumberPickerComponent.prototype.isHorizontal = /**
     * @return {?}
     */
    function () {
        return (this.buttonsOrientation !== 'v' && this.buttonsOrientation !== 'vertical');
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.onFocus = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        event.stopPropagation();
        this.isInputFocused = true;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.onBlur = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        event.stopPropagation();
        this.isInputFocused = false;
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.onMouseWheel = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.isInputFocused) {
            event.preventDefault();
            /** @type {?} */
            var wheelUp = null;
            /** @type {?} */
            var delta = null;
            if (event.deltaY) {
                delta = event.deltaY / 60;
            }
            if (event.detail) {
                delta = -event.detail / 2;
            }
            if (delta !== null) {
                wheelUp = delta > 0;
            }
            this.afterMouseWheels(wheelUp, event);
            event.stopPropagation();
        }
    };
    /**
     * @private
     * @param {?} wheelUp
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.afterMouseWheels = /**
     * @private
     * @param {?} wheelUp
     * @param {?} event
     * @return {?}
     */
    function (wheelUp, event) {
        this.onPickStarted(wheelUp);
        if (wheelUp) {
            this.onIncrease(event);
        }
        else {
            this.onDecrease(event);
        }
        this.onPickStoped(wheelUp);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.onValueChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.value > this.max) {
            this.value = this.max;
        }
        else if (this.value < this.min) {
            this.value = this.min;
        }
        if (this.parseVal(this.value)) {
            this.valueChange.emit(this.value);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.onDecrease = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        if (this.canDecrease()) {
            this.value = this.round((this.value > this.min) ? this.value -= this.step : this.value);
            this.valueChange.emit(this.value);
        }
        else {
            this.minReached.emit(true);
        }
        event.stopPropagation();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.onIncrease = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        event.preventDefault();
        if (this.canIncrease()) {
            this.value = this.round((this.value < this.max) ? this.value += this.step : this.value);
            this.valueChange.emit(this.value);
        }
        else {
            this.maxReached.emit(true);
        }
        event.stopPropagation();
    };
    /**
     * @param {?} event
     * @param {?=} increase
     * @return {?}
     */
    NumberPickerComponent.prototype.onMouseDown = /**
     * @param {?} event
     * @param {?=} increase
     * @return {?}
     */
    function (event, increase) {
        if (increase === void 0) { increase = true; }
        this.afterMouseDown(increase, event);
    };
    /**
     * @private
     * @param {?} keyCode
     * @return {?}
     */
    NumberPickerComponent.prototype.isArrowUpDown = /**
     * @private
     * @param {?} keyCode
     * @return {?}
     */
    function (keyCode) {
        return keyCode === 38 || keyCode === 40;
    };
    /**
     * @private
     * @param {?} keyCode
     * @return {?}
     */
    NumberPickerComponent.prototype.isArowUp = /**
     * @private
     * @param {?} keyCode
     * @return {?}
     */
    function (keyCode) {
        return keyCode === 38;
    };
    /**
     * @private
     * @param {?} increase
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.loopPick = /**
     * @private
     * @param {?} increase
     * @param {?} event
     * @return {?}
     */
    function (increase, event) {
        var _this = this;
        this.onPickStarted(increase);
        this.eventHolder = setTimeout((/**
         * @return {?}
         */
        function () {
            _this.countInterval = setInterval((/**
             * @return {?}
             */
            function () {
                if (increase) {
                    _this.onIncrease(event);
                }
                else {
                    _this.onDecrease(event);
                }
            }), _this.pickTimer);
        }), this.pickStartAfter);
    };
    /**
     * @param {?} event
     * @param {?=} increase
     * @return {?}
     */
    NumberPickerComponent.prototype.onMouseUp = /**
     * @param {?} event
     * @param {?=} increase
     * @return {?}
     */
    function (event, increase) {
        if (increase === void 0) { increase = true; }
        this.afterMouseUp(increase, event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.onKeyDown = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.isArrowUpDown(event.keyCode)) {
            event.preventDefault();
            if (!this.eventHolder) {
                this.loopPick(this.isArowUp(event.keyCode), event);
            }
        }
        event.stopPropagation();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.onKeyUp = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        if (this.isArrowUpDown(event.keyCode)) {
            event.preventDefault();
            this.afterPick(this.isArowUp(event.keyCode));
        }
        event.stopPropagation();
    };
    /**
     * @private
     * @param {?} up
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.afterMouseDown = /**
     * @private
     * @param {?} up
     * @param {?} event
     * @return {?}
     */
    function (up, event) {
        event.preventDefault();
        if (this.isLeftClick(event.which)) {
            this.loopPick(up, event);
        }
        event.stopPropagation();
    };
    /**
     * @private
     * @param {?} up
     * @param {?} event
     * @return {?}
     */
    NumberPickerComponent.prototype.afterMouseUp = /**
     * @private
     * @param {?} up
     * @param {?} event
     * @return {?}
     */
    function (up, event) {
        event.preventDefault();
        this.afterPick(up);
        event.stopPropagation();
    };
    /**
     * @private
     * @param {?} up
     * @return {?}
     */
    NumberPickerComponent.prototype.afterPick = /**
     * @private
     * @param {?} up
     * @return {?}
     */
    function (up) {
        this.onPickStoped(up);
        this.clearTimers();
    };
    /**
     * @private
     * @return {?}
     */
    NumberPickerComponent.prototype.clearTimers = /**
     * @private
     * @return {?}
     */
    function () {
        clearTimeout(this.eventHolder);
        clearInterval(this.countInterval);
        this.eventHolder = null;
        this.countInterval = null;
    };
    /**
     * @private
     * @param {?} up
     * @param {?} event
     * @param {?} start
     * @return {?}
     */
    NumberPickerComponent.prototype.afterArrowKeysPressed = /**
     * @private
     * @param {?} up
     * @param {?} event
     * @param {?} start
     * @return {?}
     */
    function (up, event, start) {
        if (start) {
            this.loopPick(up, event);
        }
        else {
            this.afterPick(up);
        }
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    NumberPickerComponent.prototype.parseVal = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (typeof value === 'number') {
            return value;
        }
        return parseFloat(value);
    };
    /**
     * @private
     * @param {?} step
     * @return {?}
     */
    NumberPickerComponent.prototype.getPrecision = /**
     * @private
     * @param {?} step
     * @return {?}
     */
    function (step) {
        return /\d*$/.exec(String(step))[0].length;
    };
    /**
     * @private
     * @param {?} value
     * @return {?}
     */
    NumberPickerComponent.prototype.round = /**
     * @private
     * @param {?} value
     * @return {?}
     */
    function (value) {
        return Math.round(value * Math.pow(10, this.precision)) / Math.pow(10, this.precision);
    };
    /**
     * @private
     * @return {?}
     */
    NumberPickerComponent.prototype.canIncrease = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var canIncrease = (this.value <= this.max - this.step);
        if (!canIncrease) {
            this.value = this.max;
        }
        return canIncrease;
    };
    /**
     * @private
     * @return {?}
     */
    NumberPickerComponent.prototype.canDecrease = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var canDecrease = (this.value >= this.min + this.step);
        if (!canDecrease) {
            this.value = this.min;
        }
        return canDecrease;
    };
    /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    NumberPickerComponent.prototype.onPickStarted = /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    function (increase) {
        if (increase) {
            if (this.canIncrease()) {
                this.pickStarted.emit(true);
                this.pickUpStarted.emit(true);
            }
        }
        else {
            if (this.canDecrease()) {
                this.pickStarted.emit(true);
                this.pickDownStarted.emit(true);
            }
        }
    };
    /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    NumberPickerComponent.prototype.onPickStoped = /**
     * @private
     * @param {?} increase
     * @return {?}
     */
    function (increase) {
        if (increase) {
            if (this.canIncrease()) {
                this.pickUpStoped.emit(true);
                this.pickStoped.emit(true);
            }
        }
        else {
            if (this.canDecrease()) {
                this.pickDownStoped.emit(true);
                this.pickStoped.emit(true);
            }
        }
    };
    /**
     * @private
     * @param {?} witch
     * @return {?}
     */
    NumberPickerComponent.prototype.isLeftClick = /**
     * @private
     * @param {?} witch
     * @return {?}
     */
    function (witch) {
        return witch === 1;
    };
    /**
     * @private
     * @return {?}
     */
    NumberPickerComponent.prototype.initPicker = /**
     * @private
     * @return {?}
     */
    function () {
        this.min = this.parseVal(this.min) || this.numberPickerService.min;
        this.max = this.parseVal(this.max) || this.numberPickerService.max;
        this.step = this.parseVal(this.step) || this.numberPickerService.step;
        this.value = this.parseVal(this.value) || this.numberPickerService.value;
        this.pickStartAfter = this.parseVal(this.pickStartAfter) || this.numberPickerService.pickStartAfter;
        this.pickTimer = this.parseVal(this.pickTimer) || this.numberPickerService.pickTimer;
        this.precision = this.getPrecision(this.step) || this.numberPickerService.precision;
        this.value = this.round(this.value);
        this.placeholder = this.placeholder !== undefined ? this.placeholder : '';
    };
    NumberPickerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'ng-number-picker',
                    template: "\n  <div class=\"input-group mb-3 input-{{size}} {{customClass.container}}\">\n  <!-- Horizontal decrease button orientation -->\n  <div class=\"input-group-prepend\" *ngIf=\"isHorizontal() && showDownButton\">\n    <span class=\"input-group-text decrease {{customClass.down}}\" (click)=\"onDecrease($event)\"\n    (mouseup)=\"onMouseUp($event, false)\" (mousedown)=\"onMouseDown($event, false)\">-</span>\n  </div>\n  <!-- Input prefix -->\n  <div *ngIf=\"prefix\" class=\"input-group-prepend\">\n    <span class=\"input-group-text {{customClass.prefix}}\">{{prefix}}</span>\n  </div>\n  <input type=\"number\" class=\"form-control\" name=\"input-spin-val\"\n  [(ngModel)]=\"value\"\n  [readOnly]=\"inputReadOnly\"\n  (blur)=\"onBlur($event)\"\n  (focus)=\"onFocus($event)\"\n  (mousewheel)=\"mouseWheel && onMouseWheel($event)\"\n  (keyup)=\"arrowKeys && onKeyUp($event)\"\n  (keydown)=\"arrowKeys && onKeyDown($event)\"\n  (keydown.arrowup)=\"arrowKeys && onIncrease($event)\"\n  (keydown.arrowdown)=\"arrowKeys && onDecrease($event)\"\n  (change)=\"onValueChange($event)\"\n  [placeholder]=\"placeholder\"\n  >\n  <!-- Input postfix -->\n  <div *ngIf=\"postfix\" class=\"input-group-prepend\">\n    <span class=\"input-group-text {{customClass.postfix}}\" [style.borderLeft]=\"'0'\">{{postfix}}</span>\n  </div>\n  <!-- Horizontal increase button orientation -->\n  <div class=\"input-group-prepend\" *ngIf=\"isHorizontal() && showUpButton\">\n    <span class=\"input-group-text increase {{customClass.up}}\" [style.borderLeft]=\"(!postfix) ? '0' : ''\" (click)=\"onIncrease($event)\"\n    (mouseup)=\"onMouseUp($event)\" (mousedown)=\"onMouseDown($event)\">+</span>\n  </div>\n  <!-- Vertical buttons orientation -->\n  <div class=\"input-group-append\" *ngIf=\"!isHorizontal()\">\n    <span class=\"input-group-text vertical p-0\">\n      <span *ngIf=\"showUpButton\" class=\"{{customClass.up}}\" (click)=\"onIncrease($event)\" (mouseup)=\"onMouseUp($event)\"\n      (mousedown)=\"onMouseDown($event)\">+</span>\n      <span *ngIf=\"showDownButton\" class=\"{{customClass.down}}\" (click)=\"onDecrease($event)\" (mouseup)=\"onMouseUp($event, false)\"\n      (mousedown)=\"onMouseDown($event, false)\">-</span>\n    </span>\n  </div>\n</div>\n  ",
                    styles: ["input[type=number]::-webkit-inner-spin-button,input[type=number]::-webkit-outer-spin-button{-webkit-appearance:none;margin:0}.decrease:hover,.increase:hover{-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:pointer;background-color:#d8d8d8}.increase{border-top-right-radius:3px!important;border-bottom-right-radius:3px!important}.vertical{display:flex;justify-content:center;flex-direction:column;text-align:center;-webkit-touch-callout:none;-webkit-user-select:none;-moz-user-select:none;-ms-user-select:none;user-select:none;cursor:pointer}.vertical span{line-height:15px}.input-md .vertical span,.input-sm .vertical span{padding:2px 10px}.input-lg .vertical span{padding:4px 10px}.input-xlg .vertical span{padding:7px 10px}.input-md,.input-medium{height:45px}.input-large,.input-lg{height:50px}.input-xlarge,.input-xlg{height:75px}.input-md input,.input-md span,.input-medium input,.input-medium span{font-size:22px}.input-large input,.input-large span,.input-lg input,.input-lg span{font-size:25px}.input-xlarge input,.input-xlarge span,.input-xlg input,.input-xlg span{font-size:38px}"]
                }] }
    ];
    /** @nocollapse */
    NumberPickerComponent.ctorParameters = function () { return [
        { type: NumberPickerService }
    ]; };
    NumberPickerComponent.propDecorators = {
        min: [{ type: Input }],
        max: [{ type: Input }],
        step: [{ type: Input }],
        value: [{ type: Input }],
        pickStartAfter: [{ type: Input }],
        pickTimer: [{ type: Input }],
        prefix: [{ type: Input }],
        postfix: [{ type: Input }],
        placeholder: [{ type: Input }],
        buttonsOrientation: [{ type: Input }],
        size: [{ type: Input }],
        customClass: [{ type: Input }],
        mouseWheel: [{ type: Input }],
        arrowKeys: [{ type: Input }],
        inputReadOnly: [{ type: Input }],
        showUpButton: [{ type: Input }],
        showDownButton: [{ type: Input }],
        valueChange: [{ type: Output }],
        minReached: [{ type: Output }],
        maxReached: [{ type: Output }],
        pickStarted: [{ type: Output }],
        pickStoped: [{ type: Output }],
        pickUpStarted: [{ type: Output }],
        pickUpStoped: [{ type: Output }],
        pickDownStarted: [{ type: Output }],
        pickDownStoped: [{ type: Output }]
    };
    return NumberPickerComponent;
}());
export { NumberPickerComponent };
if (false) {
    /**
     * @type {?}
     * @private
     */
    NumberPickerComponent.prototype.precision;
    /**
     * @type {?}
     * @private
     */
    NumberPickerComponent.prototype.eventHolder;
    /**
     * @type {?}
     * @private
     */
    NumberPickerComponent.prototype.countInterval;
    /**
     * @type {?}
     * @private
     */
    NumberPickerComponent.prototype.isInputFocused;
    /** @type {?} */
    NumberPickerComponent.prototype.min;
    /** @type {?} */
    NumberPickerComponent.prototype.max;
    /** @type {?} */
    NumberPickerComponent.prototype.step;
    /** @type {?} */
    NumberPickerComponent.prototype.value;
    /** @type {?} */
    NumberPickerComponent.prototype.pickStartAfter;
    /** @type {?} */
    NumberPickerComponent.prototype.pickTimer;
    /** @type {?} */
    NumberPickerComponent.prototype.prefix;
    /** @type {?} */
    NumberPickerComponent.prototype.postfix;
    /** @type {?} */
    NumberPickerComponent.prototype.placeholder;
    /** @type {?} */
    NumberPickerComponent.prototype.buttonsOrientation;
    /** @type {?} */
    NumberPickerComponent.prototype.size;
    /** @type {?} */
    NumberPickerComponent.prototype.customClass;
    /** @type {?} */
    NumberPickerComponent.prototype.mouseWheel;
    /** @type {?} */
    NumberPickerComponent.prototype.arrowKeys;
    /** @type {?} */
    NumberPickerComponent.prototype.inputReadOnly;
    /** @type {?} */
    NumberPickerComponent.prototype.showUpButton;
    /** @type {?} */
    NumberPickerComponent.prototype.showDownButton;
    /** @type {?} */
    NumberPickerComponent.prototype.valueChange;
    /** @type {?} */
    NumberPickerComponent.prototype.minReached;
    /** @type {?} */
    NumberPickerComponent.prototype.maxReached;
    /** @type {?} */
    NumberPickerComponent.prototype.pickStarted;
    /** @type {?} */
    NumberPickerComponent.prototype.pickStoped;
    /** @type {?} */
    NumberPickerComponent.prototype.pickUpStarted;
    /** @type {?} */
    NumberPickerComponent.prototype.pickUpStoped;
    /** @type {?} */
    NumberPickerComponent.prototype.pickDownStarted;
    /** @type {?} */
    NumberPickerComponent.prototype.pickDownStoped;
    /**
     * @type {?}
     * @private
     */
    NumberPickerComponent.prototype.numberPickerService;
}
