// ng-number-picker/esm2015/lib/number-picker.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class NumberPickerService {
    constructor() {
        /**
         * Min picker value
         */
        this.min = 0;
        /**
         * Max picker value
         */
        this.max = 100;
        /**
         * Pick step value
         */
        this.step = 1;
        /**
         * Delay for start picking values
         */
        this.pickStartAfter = 500;
        /**
         * Delay betweens each pick
         */
        this.pickTimer = 100;
        /**
         * value precision
         */
        this.precision = 1;
        /**
         * Inital picker value
         */
        this.value = null;
    }
}
NumberPickerService.decorators = [
    { type: Injectable }
];
if (false) {
    /**
     * Min picker value
     * @type {?}
     */
    NumberPickerService.prototype.min;
    /**
     * Max picker value
     * @type {?}
     */
    NumberPickerService.prototype.max;
    /**
     * Pick step value
     * @type {?}
     */
    NumberPickerService.prototype.step;
    /**
     * Delay for start picking values
     * @type {?}
     */
    NumberPickerService.prototype.pickStartAfter;
    /**
     * Delay betweens each pick
     * @type {?}
     */
    NumberPickerService.prototype.pickTimer;
    /**
     * value precision
     * @type {?}
     */
    NumberPickerService.prototype.precision;
    /**
     * Inital picker value
     * @type {?}
     */
    NumberPickerService.prototype.value;
}
